### PR TITLE
support jdk9 forkjoinpool maximum-pool-size

### PR DESCRIPTION
**Pull request by pjfanning**
* relates to #483 
* needs a lot more work
* open for debate on how best to proceed

### CODE REVIEW

File: actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala
Comment (He-Pin) [inline actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala:22]: Java code style for the `Constants`. And why we have such a big value? the default fork join pool is expected to not run any blocking things, and blocking code should be managed.

Comment (pjfanning) [reply actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala:22]: these are the values that java.util.concurrent.ForkJoinPool use by default. They are not public - that's why In need my own constants.

File: actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinExecutorConfigurator.scala
Comment (He-Pin) [inline actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinExecutorConfigurator.scala:97]: Maybe add a util to read the current jdk version instead.

Comment (pjfanning) [reply actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinExecutorConfigurator.scala:97]: because of the way Pekko builds the Java9+ classes, even the unit tests seem not to have this class available. It does eventually get built and appears in the jar.

File: actor/src/main/scala-jdk-9/org/apache/pekko/dispatch/PekkoJdk9ForkJoinPool.scala
Comment (He-Pin) [inline actor/src/main/scala-jdk-9/org/apache/pekko/dispatch/PekkoJdk9ForkJoinPool.scala:30]: `@InternalApi`

File: actor/src/main/resources/reference.conf
Comment (He-Pin) [inline actor/src/main/resources/reference.conf:488]: These are huge too, will cause OOM on common 4core 8GB box

Comment (pjfanning) [reply actor/src/main/resources/reference.conf:488]: please read the java.util.concurrent.ForkJoinPool code in Java 9+

this is the value hat it uses by default

Comment (migesok) [reply actor/src/main/resources/reference.conf:488]: May be it makes sense to make to turn around the configuration for FJP and make it similar to how the default global FJP in JDK is configured:
- "parallelism" calculation could stay the same, be derived from runtime CPU count
- add a new parameter called something like `max-spare-threads` (similar to "java.util.concurrent.ForkJoinPool.common.maximumSpares" system property) with 256 as the default value (again the same as the default for the global FJP)
- the resulting maximum pool size will be calculated as parallelism + max-spare-threads

Additionally I think another parameter needs to be exposed - `reject-blocking-if-pool-saturated` or something similar which governs the `saturate` predicate in the new FJP constructor - the user should have a choice between failing on blocking task once the pool limit is reached or keep piling the tasks in memory.

Does it make sense?

Comment (pjfanning) [reply actor/src/main/resources/reference.conf:488]: I'm not changing how the existing configs work - I worry far more about breaking behaviour for existing users than adding new features that only a small number of people want.

If we want new behaviour, we need new configs and to deprecate the old ones.

Please don't complicate this PR by adding extra requirements to it. Raise a new issue if you don't like the behaviour of the existing configs.

File: actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala
Comment (mdedetrich) [reply actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala]: > Java code style for the Constants

If this is a Scala source file we should be using Scala convention for constants, which would be `MaxCap` rather than `MAX_CAP`

File: actor/src/main/resources/reference.conf
Comment (migesok) [reply actor/src/main/resources/reference.conf:488]: Thank you! I see your point now about changing the existing behaviour. I just want to suggest that instead of exposing `maximum-pool-size` directly, it might be better to expose the spare thread count as JDK devs did for the global pool. The default value could be similar to the current default (effectively no limit at all).
Given that parallelism is calculated programmatically in the current Akka FJP configurator, combining it with a `maximum-pool-size` when the values are close might be inconvenient. Usually when you look at it, you think in the terms of "how many parallel blocking operations I can afford, additionally to the normal parallelism level".

I can contribute exposing `saturate` config in a separate PR later if this one goes through.

**Comment by He-Pin** [comment]
We are using Pekko now, and all of our nodes are running on Java 21, I would like to see this feature be shipped in 1.1.x.

**Comment by mdedetrich** [comment]
> We are using Pekko now, and all of our nodes are running on Java 21, I would like to see this feature be shipped in 1.1.x.

I don't think this is possible to do properly without multi-release-jar which is [hitting problems](https://github.com/sbt/sbt-multi-release-jar/issues/24#issuecomment-1891156070)

**Comment by He-Pin** [comment]
> > We are using Pekko now, and all of our nodes are running on Java 21, I would like to see this feature be shipped in 1.1.x.
> 
> I don't think this is possible to do properly without multi-release-jar which is [hitting problems](https://github.com/sbt/sbt-multi-release-jar/issues/24#issuecomment-1891156070)

Why? Netty can support multi jdk just with some features tags.

If that method is not called, no problem, we can built a release with jdk 11.

**Comment by pjfanning** [comment]
This code can be written using Java reflection. The PR as it stands uses Java reflection.

**Comment by He-Pin** [comment]
I think we should go with methodHanfle, this can be done whiteout the multi release jar support.

**Comment by He-Pin** [comment]
> This code can be written using Java reflection. The PR as it stands uses Java reflection.

Better with method handle, on Java 17+, we need an addition add opens, the less reflection the better,we should move on.

**Comment by pjfanning** [comment]
> > This code can be written using Java reflection. The PR as it stands uses Java reflection.
> 
> Better with method handle, on Java 17+, we need an addition add opens, the less reflections the better,we should move on prepare for future Java releases.

I've changed the PR to use a MethodHandle.

**Comment by pjfanning** [comment]
@mdedetrich He Pin wants this in v1.1.0. I'm neutral on this. Is it ok to use MethodHandles like this and rewrite later when we have Multi-Release jar support?

**Comment by He-Pin** [comment]
@pjfanning I want this because it will dress some issue especially when user is blocking the actors.
image

We have something like this in our production code with Virtual Thread too, If you want to defer this to 1.2.x/ 2.0.x that's fine, but this will be a killer feature in 1.1.x.

File: actor/src/main/resources/reference.conf
Comment (He-Pin) [reply actor/src/main/resources/reference.conf:488]: The virtual thread is setting to `256` by default if the `maxPoolSizeValue` is not set, but yes, we can not change it and user can set to a limit match their systeam.

Comment (pjfanning) [reply actor/src/main/resources/reference.conf:488]: I'm not sure what these comments mean. If someone wants changes, can they provide actual code. I can interpret code more easily than abstract comments.

I have read and re-read the ForkJoinPool code and if we use the old Java constructor (as opposed to the new one added in Java 9) - that the old constructor calls the new constructor and passes the value of MAX_CAP for the maxPoolSizeValue. MAX_CAP=0x7fff which is 32767.

I will make changes if I am wrong but please use code and code links to show me where I am going wrong.

Comment (He-Pin) [reply actor/src/main/resources/reference.conf:488]: 1. I mean your current code is right, 
2. we can not change it to a smaller number,eg 256
3. user can change it  if they like.

File: actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala
Comment (He-Pin) [inline actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala:22]: DefaultKeepAliveInMiilis

File: actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinExecutorConfigurator.scala
Comment (He-Pin) [reply actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinExecutorConfigurator.scala:98]:
```scala
private lazy val pekkoJdk9ForkJoinPoolHandleOpt: Option[MethodHandle] = {
     if (JavaVersion.majorVersion >= 9) {
      pekkoJdk9ForkJoinPoolClassOpt.map { clz =>
        val methodHandleLookup = MethodHandles.lookup()
        val mt = MethodType.methodType(classOf[Unit], classOf[Int],
          classOf[ForkJoinPool.ForkJoinWorkerThreadFactory],
          classOf[Int], classOf[Thread.UncaughtExceptionHandler], classOf[Boolean])
        methodHandleLookup.findConstructor(clz, mt)
      }} else None
    }
```

**Review by He-Pin — APPROVED** [review]
lgtm

**Comment by pjfanning** [comment]
> lgtm at the first round , just a little nitpick , I think we should enable run test on JDK 21 too, if the code is target to higher JDK version.

We do test with java 21. Check the nightly builds

**Review by He-Pin — APPROVED** [review]
Lgtm

File: actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala
Comment (pjfanning) [reply actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala]: done

File: actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinExecutorConfigurator.scala
Comment (pjfanning) [reply actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinExecutorConfigurator.scala:100]: I added the java-version check because it happens inside the lazy val calculation - see we won't check this over and over.

File: actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala
Comment (He-Pin) [reply actor/src/main/scala/org/apache/pekko/dispatch/ForkJoinPoolConstants.scala]: Thank you, but Take care, time to sleep 😴 💤 😪

**Comment by He-Pin** [comment]
The code is well done, and the last part should be update the document, that can come up later too.

**Comment by mdedetrich** [comment]
I would be more comfortable putting this in 1.1.0-M2 or maybe 1.2.0 as we are already overloaded with changes and we have just created a regression with JDK

**Comment by pjfanning** [comment]
@He-Pin out of interest, is there a reason why you can't create a custom dispatcher that supports java9+ fork-join-pool? Such an approach would work today, no reason to wait for a Pekko 1.1 release.

**Comment by He-Pin** [comment]
> @He-Pin out of interest, is there a reason why you can't create a custom dispatcher that supports java9+ fork-join-pool? Such an approach would work today, no reason to wait for a Pekko 1.1 release.

1. Custom dispatcher is much change for most crud available Java boies/girls.
2. Out of box solution is better, my systems is full async, but not all.
3. This is a demand from the pekko community, not by me.

I'm find to postpone this to later version eg 1.1.0-m2 or m3 , and I can help to test it.

**Comment by mdedetrich** [comment]
@He-Pin Argument would be a lot more convincing if creating a JDK 9 `ForkJoinPool` would be impossible which is not the case. Although having it default to the better version of JDK9+ would be nice, its also not a deal breaker and we can also document this as a stop gap.

The reason I am hesitant is that the current Pekko build with all of the JDK9 classes, multi-jvm etc etc is a bit of a mess, note that I am _**NOT**_ blaming anyone specifically here, there are legitimate reasons for this mess, i.e. Akka was a really old sbt project that had to implement various workarounds which are no longer needed but those workarounds were still left in the code and then there is the case of working around even current day sbt limitations which is what is causing the sbt-osgi problems.

In short I would like to start devoting actual effort to cleaning up the sbt build which is something that I have already started doing rather than dumping additional features that involve non trivial changes to the build which then end up causing regressions (either at the start or later down the track due to how the build interactions work). sbt-osgi is a good example of this, I took initiate/responsibility to get it working and as you can see we are still getting problems 3-4 months down the track.

We need to be more strategic/intelligent about what features we push and for this feature specifically I would prefer if we actually put it on hold until I figure out and then stabilize all of the sbt-multi-jar/sbt-multi-jvm/sbt-osgi interactions.

**Comment by He-Pin** [comment]
Yes, I want to polish the toolchain too. WDYT, a document still needs to be done to ship this. but the code should be mostly ready.
Let's delay it to 1.2.x. That will make our life easier because we must fix the toolchain first.

**Comment by mdedetrich** [comment]
If 1.1.0-M2 happens and it takes a long time because of other reasons then it can still land for 1.1.0 but to make it clear, we shouldn't delay 1.1.0 just for this feature.

I would also prefer to implement this properly using multi-jar-release rather than reflection.

**Comment by He-Pin** [comment]
That's true, I expect 1.1.0-M1 to be released soon, and I will not submit any feature pr before the 1.1.x branch cut.

**Comment by pjfanning** [comment]
@He-Pin would you support having this in v1.1.0 ?

**Comment by He-Pin** [comment]
> @He-Pin would you support having this in v1.1.0 ?

Yes, I would like to have it ,even in m1.

**Comment by pjfanning** [comment]
@mdedetrich can we merge this? I can rewrite this in another release to use Multi-Release jar support. The impl is private and controlled solely by config so it is easy to just completely rewrite it. We can document that the current support is experimental.

**Review by mdedetrich — APPROVED** [review]
lgtm, as long as we can improve on it later without breaking backwards compatibility

**Review by Roiocam — APPROVED** [review]
LGTM

**Pull request merged**
